pkg/statistics/handle/util/auto_analyze_proc_id_generator.go

```diff
@@ -14,27 +14,111 @@
 
 package util
 
+import (
+	"sync"
+
+	"github.com/pingcap/tidb/pkg/sessionctx/sysproctrack"
+	"golang.org/x/exp/maps"
+)
+
 // AutoAnalyzeProcIDGenerator is used to generate auto analyze proc ID.
 type AutoAnalyzeProcIDGenerator interface {
 	// AutoAnalyzeProcID generates an analyze ID.
 	AutoAnalyzeProcID() uint64
+	ReleaseAutoAnalyzeProcID(uint64)
 }
 
 var _ AutoAnalyzeProcIDGenerator = (*generator)(nil)
 
 type generator struct {
 	// autoAnalyzeProcIDGetter is used to generate auto analyze ID.
-	autoAnalyzeProcIDGetter func() uint64
+	autoAnalyzeProcIDGetter  func() uint64
+	releaseAutoAnalyzeProcID func(uint64)
 }
 
 // NewGenerator creates a new Generator.
-func NewGenerator(autoAnalyzeProcIDGetter func() uint64) AutoAnalyzeProcIDGenerator {
+func NewGenerator(autoAnalyzeProcIDGetter func() uint64, releaseAutoAnalyzeProcID func(uint64)) AutoAnalyzeProcIDGenerator {
 	return &generator{
-		autoAnalyzeProcIDGetter: autoAnalyzeProcIDGetter,
+		autoAnalyzeProcIDGetter:  autoAnalyzeProcIDGetter,
+		releaseAutoAnalyzeProcID: releaseAutoAnalyzeProcID,
 	}
 }
 
 // AutoAnalyzeProcID implements AutoAnalyzeProcIDGenerator.
 func (g *generator) AutoAnalyzeProcID() uint64 {
 	return g.autoAnalyzeProcIDGetter()
 }
+
+// ReleaseAutoAnalyzeProcID implements AutoAnalyzeProcIDGenerator.
+func (g *generator) ReleaseAutoAnalyzeProcID(id uint64) {
+	g.releaseAutoAnalyzeProcID(id)
+}
+
+// GlobalAutoAnalyzeProcessList is used to track the auto analyze process.
+var GlobalAutoAnalyzeProcessList = newGlobalAutoAnalyzeProcessList()
+
+type globalAutoAnalyzeProcessList struct {
+	processes map[uint64]struct{}
```

Comment:
We use global thread-safe map to record the running analyze task.

Reply:
why not use a `sync.map`

Reply:
The Map type is optimized for two common use cases: (1) when the entry for a given key is only ever written once but read many times, as in caches that only grow, or (2) when multiple goroutines read, write, and overwrite entries for disjoint sets of keys. In these two cases, use of a Map may significantly reduce lock contention compared to a Go map paired with a separate Mutex or RWMutex. 

so two common use cases are not the same as us.